### PR TITLE
refactor(catchError): simplify implementation

### PR DESCRIPTION
**Pull request by josepot**
**Description:**

After I saw #5665 I decided to have a look at the implementation of `catchError`, and when I did I thought that it could be simplified. So, I wrote a simpler implementation and with this implementation all the tests pass and ~on top of that it also fixes the issue described on #5665~ 

~I guess that I should also write a test that reproduces the issue described on #5665, but I couldn't come up with a good description for a test like that 🤔. Maybe tomorrow I will be more inspired 😄~

**Related issue (if exists):**
#5665 

## EDIT

It turns out that the the issue described on #5665 was fixed by #5627. So, this PR doesn't fix anything. However, it does simplify the code quite a bit IMO.

### CODE REVIEW

File: src/internal/operators/catchError.ts [hunk 1]
```diff
@@ -111,44 +111,24 @@ export function catchError<T, O extends ObservableInput<any>>(
...
-        innerSub = null;
-        subscription.add(handledResult!.subscribe(subscriber));
-      } else {
-        subscription.add(innerSub);
-      }
-
-      return subscription;
+      return () => {
+        subscription!.unsubscribe();
+        subscription = sourceSubscription = null;
```
[the remark this answers is not in the document]
Comment (benlesh) [reply src/internal/operators/catchError.ts]: Subscriptions are often subscribers, and historically, our subscribers are these OOP monstrosities that have a lot of state living in properties. There's this practice throughout the library mostly to make sure things that are no longer being used get orphaned for GC. That said, some of it is _probably_ superstition. I haven't examined this example too closely.

File: src/internal/operators/catchError.ts [hunk 2]
```diff
@@ -111,44 +111,22 @@ export function catchError<T, O extends ObservableInput<any>>(
...
-            if (innerSub) {
-              innerSub.unsubscribe();
-              innerSub = null;
-              subscription.add(handledResult.subscribe(subscriber));
-            } else {
-              syncUnsub = true;
-            }
+          try {
+            const nextSource = from(selector(err, catchError(selector)(source)));
+            subscription = nextSource.subscribe(subscriber);
```
[the remark this answers is not in the document]
Comment (benlesh) [reply src/internal/operators/catchError.ts]: I actually hate that test with the `concat(defer, defer, defer)`. It should probably be just a plain observable with a loop, similar to "firehose" tests @cartant recently added.

File: src/internal/operators/catchError.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (cartant) [reply src/internal/operators/catchError.ts]: @josepot Regarding the nulling out, the `shareReplay` refactor/simplification you did a while ago touched on code where the nulling was important - as the subscription was **not held on a per-subscriber** context. Nulling out is far less of an issue - as Ben has mentioned - when the reference is per-subscriber.

File: src/internal/operators/catchError.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (josepot) [reply src/internal/operators/catchError.ts]: > I actually hate that test with the concat(defer, defer, defer). It should probably be just a plain observable with a loop, similar to "firehose" tests @cartant recently added.

Totally agree. In fact I'm pretty sure that test is redundant because unless I'm mistaken the "firehose" test is already covering that case, right?

File: src/internal/operators/catchError.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (josepot) [reply src/internal/operators/catchError.ts]: Thank you both @benlesh and @cartant for taking the time to answer this question, and I apologize if I framed the question in a way that it sounded like I was being judgemental. I understand that it's better to be cautious than to be sorry.

> Nulling out is far less of an issue - as Ben has mentioned - when the reference is per-subscriber.

That's what I thought, and yep, in the case of the `shareReplay` I could see why the cleanup was important in there. Once again: thank you both for the detailed explanation, much much appreciated.

**Comment by josepot** [comment]
> I don't think we can merge this.
> 
> The problem with the new implementation is that, in the synchronous case, `finalize` will now happen _after_ everything else... 

> I'm not sure if there's a test that covers this though, which might be a better add. Along with the other test you've added.

Nice! Thanks for spotting that @benlesh , I will add a test for that.

**Comment by josepot** [comment]
I've addressed the comments. There are 2 commits in this PR:

- The first one adds 2 new tests, removes a test that seems to be redundant and removes a stale `TODO` comment.
- The second one is a refactor of the `catchError` operator that IMO makes the implementation a bit simpler/clearer

It goes without saying that I don't mind at all dropping the second commit if you prefer the current implantation. It's just a suggestion: if you like it "yay!", if you don't I totally get it, of course 🙂 .

Also, maybe it would be better to create a separate PR just for the improvements of the tests?

File: src/internal/operators/catchError.ts [hunk 3]
```diff
@@ -111,44 +111,29 @@ export function catchError<T, O extends ObservableInput<any>>(
...
-            if (innerSub) {
-              innerSub.unsubscribe();
-              innerSub = null;
-              subscription.add(handledResult.subscribe(subscriber));
-            } else {
-              syncUnsub = true;
-            }
+          try {
+            nextSource = from(selector(err, catchError(selector)(source)));
+            tryToSwapSubscription();
```
Comment (benlesh) [inline src/internal/operators/catchError.ts:127]: There's another interesting semantic difference here. Any error that occurs in `tryToSwapSubscription()` will be caught and sent down the `error` channel... I'm not sure what this could mean. In the event that someone is using the deprecated sync error throwing (set in by `config`), it's possible that they have a synchronous error thrown during this subscription... I'm not sure what the behavior in this case here would be.. (And I'm equally unsure what the behavior would be in the previous code).

I guess I would expect it to send down the error channel as you do here, so this is probably correct. It's just something to think about.

Probably nothing we should be concerned with, I suppose. But I wanted to note the concern.

File: src/internal/operators/catchError.ts [hunk 4]
```diff
@@ -111,44 +111,29 @@ export function catchError<T, O extends ObservableInput<any>>(
...
-        innerSub.unsubscribe();
-        innerSub = null;
-        subscription.add(handledResult!.subscribe(subscriber));
-      } else {
-        subscription.add(innerSub);
-      }
+      tryToSwapSubscription();
 
-      return subscription;
+      return () => subscription.unsubscribe();
```
Comment (benlesh) [inline src/internal/operators/catchError.ts:136]: You should just return the `subscription` here, TBH. In master, we only just recently stopped wrapping these sorts of functions in Subscriptions. But either way, you're wrapping a subscription in a function, which isn't necessary.

File: spec/operators/catchError-spec.ts [hunk 5]
```diff
@@ -477,4 +451,57 @@ describe('catchError operator', () => {
...
+    const subscription = throwError(new Error('Some error')).pipe(
+      catchError(() => testCleanup$),
+    ).subscribe()
+    expect(clean).to.equal(false);
+
+    subscription.unsubscribe()
+    expect(clean).to.equal(true);
+  });
+
+  it('should not alter the order of the teardown logic of the subscription chain', () => {
```
Comment (benlesh) [inline spec/operators/catchError-spec.ts:469]: This test is a little too clever.

I'd prefer that things were more clearly named. For example, you're leveraging `i` in that for loop, but from the name it's not clear what the intended use of `i` is, or why it lives outside of the `for` declaration. You have to study the code to realize its part of the test, because you're throwing if you "make it through the loop". That test is probably unnecessary BTW.

Basically just an `of(1, 2, 3, 4, 5, 6, 7)` or something would probably be fine.

File: src/internal/operators/catchError.ts (hunk 4 above)
Comment (josepot) [reply src/internal/operators/catchError.ts:136]: Right! I thought that it was necessary because the subscription is being reassigned on the `tryToSwapSubscription` function. However, now that I have a closer look at it, that reassignment is not necessary. Nice!

File: spec/operators/catchError-spec.ts (hunk 5 above)
Comment (josepot) [reply spec/operators/catchError-spec.ts:469]: Yep, I got a bit lazy and I just made a test out of [your comment](https://github.com/ReactiveX/rxjs/pull/5666#pullrequestreview-475053120) :see_no_evil: 

I will improve that, for sure. Thanks!

> That test is probably unnecessary BTW.

I don't know, I think that it doesn't hurt, with my previous implementation all the tests were passing, but this test would have failed... Let me try to make it more explicit and understandable.

**Comment by josepot** [comment]
> Thanks! We're getting closer.

Once again, thank you both for the reviews. I'm learning a lot!

**Comment by benlesh** [comment]
LGTM... I'm going to bump this to @cartant since he reviewed this once earlier as well.

**Comment by cartant** [comment]
I'm going to block this, for the moment. I started reviewing it and AFAICT, there is a problem with [this bit](https://github.com/ReactiveX/rxjs/blob/e42249a532a9998156f3a783413d8953c9de7c56/src/internal/operators/catchError.ts#L119). I cannot see how `subscribe` can be safely called with no regard to the `Subscription` that has already been returned. ATM, I'm inclined to say that this implementation is potentially more confusing than the previous implementation. I'll have a think about it and will unblock it after I've reviewed it. ATM, I am not at all sure about the implementation or whether it makes things clearer.

**Comment by josepot** [comment]
@cartant 

> ATM, I'm inclined to say that this implementation is potentially more confusing than the previous implementation.

You are right. It is more confusing. I agree, my apologies.

> I cannot see how subscribe can be safely called with no regard to the Subscription that has already been returned

I made that change when I was addressing [this comment](https://github.com/ReactiveX/rxjs/pull/5666#discussion_r478734781). I remember that when I saw that comment I thought: "did Ben realize that I'm swapping the subscription? He probably missed that, because if I just return `subscription` then the tests will fail..." So, just to be sure, I tried doing that and I was very surprised to see that the tests were passing. Then I tried to rationalize why that was happening. However, to be completely honest, I don't think that I fully understood why the tests were passing.

In retrospective I should have asked for further clarification on that comment, instead.

I just put the code back to what it was before I addressed that comment. I think that implementation is a bit simpler and clearer, but this is something very subjective. So, if you don't see it like that, I totally understand it.

**Review by benlesh — CHANGES_REQUESTED** [review]
No changes requested... just removing approval so I doesn't show up in my merge list.

**Comment by josepot** [comment]
No worries at all. I'm closing this. Thank you both!

If you want me to open a different PR that includes just the commit with the tests, let me know it.

**Comment by benlesh** [comment]
Much respect, @josepot. Really. ❤️
